File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/PgpApiRepository.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node;

import com.flowcrypt.email.api.retrofit.request.node.DecryptMsgRequest;
import com.flowcrypt.email.api.retrofit.response.model.node.NodeKeyDetails;
import com.flowcrypt.email.api.retrofit.response.node.NodeResponseWrapper;

import androidx.lifecycle.MutableLiveData;

/**
 * It's an entry point of all requests to work with PGP actions.
 *
 * @author Denis Bondarenko
 * Date: 2/15/19
 * Time: 10:25 AM
 * E-mail: DenBond7@gmail.com
 */
public interface PgpApiRepository {
  /**
   * Parse the given raw string and fetch a list of {@link NodeKeyDetails}.
   *
   * @param requestCode The unique request code for identify the current action.
   * @param liveData    An instance of {@link MutableLiveData} which will be used for the result delivering.
   * @param raw         The raw string which can take one key or many keys,
   *                    it can be private or public keys, it can be armored or binary.. doesn't matter.
   */
  void fetchKeyDetails(int requestCode, MutableLiveData<NodeResponseWrapper> liveData, String raw);

  /**
   * Parse the given raw MIME message and decrypt some parts if needed.
   *
   * @param requestCode The unique request code for identify the current action.
   * @param liveData    An instance of {@link MutableLiveData} which will be used for the result delivering.
   */
  void parseAndDecryptMsg(int requestCode, MutableLiveData<NodeResponseWrapper> liveData,
                          DecryptMsgRequest decryptMsgRequest);
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/PgpApiRepository.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node

import com.flowcrypt.email.api.retrofit.base.BaseApiRepository
import com.flowcrypt.email.api.retrofit.request.node.ParseDecryptMsgRequest
import com.flowcrypt.email.api.retrofit.response.base.Result
import com.flowcrypt.email.api.retrofit.response.node.ParseDecryptedMsgResult

/**
 * It's an entry point of all requests to work with PGP actions.
 *
 * @author Denis Bondarenko
 * Date: 2/15/19
 * Time: 10:25 AM
 * E-mail: DenBond7@gmail.com
 */
//todo-denbond7 need to review this class. Maybe some things can be migrated to use coroutines
interface PgpApiRepository : BaseApiRepository {
  /**
   * Parse the given raw MIME message and decrypt some parts if needed.
   *
   * @param requestCode A unique request code for identify the current action.
   * @param request     An instance of [ParseDecryptMsgRequest] which contains information about a message.
   */
  suspend fun parseDecryptMsg(
    requestCode: Int = 0,
    request: ParseDecryptMsgRequest
  ): Result<ParseDecryptedMsgResult?>
}
